File: MGLLMDemoVB/ChatBotLogic.vb
Imports System.IO
Imports System.Net.Http
Imports System.Net.Http.Json
Imports System.Text
Imports System.Text.Json
Imports System.Text.Json.Serialization
Imports Microsoft.VisualBasic
Imports YamlDotNet.Serialization
Imports YamlDotNet.Serialization.NamingConventions

Public NotInheritable Class Message
    Public Property Role As String
    Public Property Content As String
End Class

Public NotInheritable Class DeepSeekRequest
    Public Property Model As String
    Public Property Messages As List(Of Message)
    Public Property Temperature As Double
    Public Property MaxTokens As Integer
    Public Property FrequencyPenalty As Double
    Public Property PresencePenalty As Double
End Class

Public NotInheritable Class Choice
    Public Property Message As Message
End Class

Public NotInheritable Class DeepSeekResponse
    Public Property Choices As List(Of Choice)
End Class

Public NotInheritable Class VoiceConfig
    <JsonPropertyName("assistant")> Public Property Assistant As String
    <JsonPropertyName("user")> Public Property User As String
End Class

Public NotInheritable Class ModelParams
    <JsonPropertyName("temperature")> Public Property Temperature As Double
    <JsonPropertyName("max_tokens")> Public Property MaxTokens As Integer
    <JsonPropertyName("frequency_penalty")> Public Property FrequencyPenalty As Double
    <JsonPropertyName("presence_penalty")> Public Property PresencePenalty As Double
End Class

Public NotInheritable Class AppConfig
    <JsonPropertyName("api_key")> Public Property ApiKey As String
    <JsonPropertyName("system_prompt")> Public Property SystemPrompt As List(Of String)
    <JsonPropertyName("model_params")> Public Property ModelParams As ModelParams
End Class

Public NotInheritable Class ChatLogEntry
    Public Property Index As Integer
    Public Property Timestamp As String
    Public Property UserMessage As String
    Public Property ChatbotResponse As String
End Class

Public Module ChatBotLogic
    Private _currentIndex As Integer = 0

    Private Property Config As AppConfig

    Public Sub LoadAppConfig()
        Dim configPath = Path.Combine(Directory.GetCurrentDirectory(), "mgllm-config.json")
        If File.Exists(configPath) Then
            Dim jsonContent = File.ReadAllText(configPath)
            Config = JsonSerializer.Deserialize(Of AppConfig)(jsonContent)
        Else
            Config = CreateDefaultAppConfig()
            Try
                Dim prompt = "Please enter your DeepSeek API key. This will not be saved to disk by default.

If you leave this blank the application will still run but API requests will fail."
                Dim entered = InputBox(prompt, "DeepSeek API Key", "")
                If Not String.IsNullOrWhiteSpace(entered) Then Config.ApiKey = entered.Trim()
            Catch ex As Exception
                MsgBox(ex.ToString(), Title:=ex.GetType().Name)
            End Try
        End If
        ApplyConfigFallbacks()
    End Sub

    Private Function CreateDefaultAppConfig() As AppConfig
        Return New AppConfig With {
            .ApiKey = String.Empty,
            .SystemPrompt = New List(Of String),
            .ModelParams = New ModelParams With {
                .Temperature = 0.7,
                .MaxTokens = 500,
                .FrequencyPenalty = 0.3,
                .PresencePenalty = 0.3
            }
        }
    End Function

    Private Sub ApplyConfigFallbacks()
        Dim defaultAppConfig = CreateDefaultAppConfig()
        If Config Is Nothing Then
            Config = defaultAppConfig
            Exit Sub
        End If
        With Config
            If .SystemPrompt Is Nothing Then .SystemPrompt = defaultAppConfig.SystemPrompt
            If .ModelParams Is Nothing Then .ModelParams = defaultAppConfig.ModelParams
        End With
    End Sub

    Private Const DEEPSEEK_API_URL As String = "https://api.deepseek.com/chat/completions"

    Private ReadOnly Property ChatLogYamlFilePath As String
        Get
            Return Path.Combine(Directory.GetCurrentDirectory(), "mgllm-chatlog.yml")
        End Get
    End Property

    Public Async Function GetDeepSeekResponseAsync _
            (userMessage As String, Optional conversationHistory As List(Of Message) = Nothing) _
            As Task(Of String)
        Using httpClient As New HttpClient
            httpClient.DefaultRequestHeaders.Add("Authorization", $"Bearer {Config.ApiKey}")

            Dim messages As New List(Of Message) From {
                New Message With {
                    .Role = "system",
                    .Content = String.Join(vbCrLf, Config.SystemPrompt)
                }
            }

            If conversationHistory IsNot Nothing Then
                For Each msg As Message In conversationHistory
                    messages.Add(New Message With {.Role = msg.Role, .Content = msg.Content})
                Next msg
            End If
            messages.Add(New Message With {.Role = "user", .Content = userMessage})

            Dim requestBody As New DeepSeekRequest With {
                .Model = "deepseek-v4-flash",
                .Messages = messages,
                .Temperature = Config.ModelParams.Temperature,
                .MaxTokens = Config.ModelParams.MaxTokens,
                .FrequencyPenalty = Config.ModelParams.FrequencyPenalty,
                .PresencePenalty = Config.ModelParams.PresencePenalty
            }

            Dim options As New JsonSerializerOptions With {
                .PropertyNamingPolicy = JsonNamingPolicy.CamelCase
            }
            Dim jsonContent As String = JsonSerializer.Serialize(requestBody, options)
            Dim content As New StringContent(jsonContent, Encoding.UTF8, "application/json")

            Try
                Dim response = Await httpClient.PostAsync(DEEPSEEK_API_URL, content)
                Dim responseContent = Await response.Content.ReadAsStringAsync()

                If response.IsSuccessStatusCode Then
                    Dim jsonResponse =
                        JsonSerializer.Deserialize(Of DeepSeekResponse)(responseContent, options)
                    Return jsonResponse.Choices(0).Message.Content
                Else
                    Return $"API Error: {response.StatusCode} - {responseContent}"
                End If
            Catch ex As Exception
                Return $"Request failed: {ex.Message}"
            End Try
        End Using
    End Function

    Public ReadOnly Property NextConversationIndex As Integer
        Get
            _currentIndex += 1
            Return _currentIndex
        End Get
    End Property

    Public Sub SaveChatLog(userMessage As String, botResponse As String, index As Integer)
        Dim timestamp As String = Date.Now.ToString("yyyy-MM-dd HH:mm:ss")
        Dim namingConvention = CamelCaseNamingConvention.Instance
        Dim entries As New List(Of ChatLogEntry)

        If File.Exists(ChatLogYamlFilePath) Then
            Dim yamlContent As String = File.ReadAllText(ChatLogYamlFilePath)
            Dim deserializer =
                (New DeserializerBuilder).WithNamingConvention(namingConvention).Build()
            entries = deserializer.Deserialize(Of List(Of ChatLogEntry))(yamlContent)
        End If
        If entries Is Nothing Then entries = New List(Of ChatLogEntry)
        entries.Add(New ChatLogEntry With {
            .Index = index,
            .Timestamp = timestamp,
            .UserMessage = userMessage,
            .ChatbotResponse = botResponse
        })

        Dim serializer = (New SerializerBuilder).WithNamingConvention(namingConvention).Build()
        File.WriteAllText(ChatLogYamlFilePath, serializer.Serialize(entries), Encoding.UTF8)
    End Sub
End Module

File: MGLLMDemoCSharp/ChatBotLogic.cs
using System.Text;
using System.Text.Json;
using System.Text.Json.Serialization;
using YamlDotNet.Serialization;
using YamlDotNet.Serialization.NamingConventions;
using Microsoft.VisualBasic;

namespace MGLLMDemoCSharp;

public sealed class Message
{
    public string? Role { get; set; }
    public string? Content { get; set; }
}

public sealed class DeepSeekRequest
{
    public string? Model { get; set; }
    public List<Message>? Messages { get; set; }
    public double Temperature { get; set; }
    public int MaxTokens { get; set; }
    public double FrequencyPenalty { get; set; }
    public double PresencePenalty { get; set; }
}

public sealed class Choice
{
    public Message? Message { get; set; }
}

public sealed class DeepSeekResponse
{
    public List<Choice>? Choices { get; set; }
}

public sealed class VoiceConfig
{
    [JsonPropertyName("assistant")]
    public string? Assistant { get; set; }
    [JsonPropertyName("user")]
    public string? User { get; set; }
}

public sealed class ModelParams
{
    [JsonPropertyName("temperature")]
    public double Temperature { get; set; }
    [JsonPropertyName("max_tokens")]
    public int MaxTokens { get; set; }
    [JsonPropertyName("frequency_penalty")]
    public double FrequencyPenalty { get; set; }
    [JsonPropertyName("presence_penalty")]
    public double PresencePenalty { get; set; }
}

public sealed class AppConfig
{
    [JsonPropertyName("api_key")]
    public string? ApiKey { get; set; }
    [JsonPropertyName("system_prompt")]
    public List<string>? SystemPrompt { get; set; }
    [JsonPropertyName("model_params")]
    public ModelParams? ModelParams { get; set; }
}

public sealed class ChatLogEntry
{
    public int Index { get; set; }
    public string? Timestamp { get; set; }
    public string? UserMessage { get; set; }
    public string? ChatbotResponse { get; set; }
}

public static class ChatBotLogic
{
    private static int _currentIndex = 0;

    private static AppConfig? Config { get; set; }

    public static void LoadAppConfig()
    {
        string configPath = Path.Combine(Directory.GetCurrentDirectory(), "mgllm-config.json");
        if (File.Exists(configPath))
        {
            string jsonContent = File.ReadAllText(configPath);
            Config = JsonSerializer.Deserialize<AppConfig>(jsonContent);
        }
        else
        {
            Config = CreateDefaultAppConfig();
            try
            {
                string prompt = @"Please enter your DeepSeek API key. This will not be saved to disk by default.

If you leave this blank the application will still run but API requests will fail.";
                string entered = Interaction.InputBox(prompt, "DeepSeek API Key", "");
                if (!string.IsNullOrWhiteSpace(entered))
                    Config.ApiKey = entered.Trim();
            }
            catch (Exception ex)
            {
                Interaction.MsgBox(ex.ToString(), Title: ex.GetType().Name);
            }
        }
        ApplyConfigFallbacks();
    }

    private static AppConfig CreateDefaultAppConfig()
    {
        return new AppConfig()
        {
            ApiKey = string.Empty,
            SystemPrompt = [],
            ModelParams = new ModelParams()
            {
                Temperature = 0.7,
                MaxTokens = 500,
                FrequencyPenalty = 0.3,
                PresencePenalty = 0.3
            }
        };
    }

    private static void ApplyConfigFallbacks()
    {
        var defaultAppConfig = CreateDefaultAppConfig();
        if (Config is null)
        {
            Config = defaultAppConfig;
            return;
        }
        Config.SystemPrompt ??= defaultAppConfig.SystemPrompt;
        Config.ModelParams ??= defaultAppConfig.ModelParams;
    }

    private const string DeepSeekApiUrl = "https://api.deepseek.com/chat/completions";

    private static string ChatLogYamlFilePath =>
        Path.Combine(Directory.GetCurrentDirectory(), "mgllm-chatlog.yml");

    public async static Task<string> GetDeepSeekResponseAsync(string userMessage, List<Message> conversationHistory = null!)
    {
        using HttpClient httpClient = new();
        httpClient.DefaultRequestHeaders.Add("Authorization", $"Bearer {Config!.ApiKey}");

        List<Message> messages =
        [
            new()
            {
                Role = "system",
                Content = string.Join("\n", Config.SystemPrompt!)
            }
        ];

        if (conversationHistory is not null)
        {
            foreach (Message msg in conversationHistory)
                messages.Add(new Message() { Role = msg.Role, Content = msg.Content });
        }
        messages.Add(new Message() { Role = "user", Content = userMessage });

        DeepSeekRequest requestBody = new()
        {
            Model = "deepseek-v4-flash",
            Messages = messages,
            Temperature = Config.ModelParams!.Temperature,
            MaxTokens = Config.ModelParams.MaxTokens,
            FrequencyPenalty = Config.ModelParams.FrequencyPenalty,
            PresencePenalty = Config.ModelParams.PresencePenalty
        };

        JsonSerializerOptions options = new() { PropertyNamingPolicy = JsonNamingPolicy.CamelCase };
        string jsonContent = JsonSerializer.Serialize(requestBody, options);
        StringContent content = new(jsonContent, Encoding.UTF8, "application/json");

        try
        {
            var response = await httpClient.PostAsync(DeepSeekApiUrl, content);
            string responseContent = await response.Content.ReadAsStringAsync();

            if (response.IsSuccessStatusCode)
            {
                var jsonResponse = JsonSerializer.Deserialize<DeepSeekResponse>(responseContent, options);
                return jsonResponse!.Choices![0].Message!.Content!;
            }
            else
            {
                return $"API Error: {response.StatusCode} - {responseContent}";
            }
        }
        catch (Exception ex)
        {
            return $"Request failed: {ex.Message}";
        }
    }

    public static int NextConversationIndex
    {
        get
        {
            _currentIndex += 1;
            return _currentIndex;
        }
    }

    public static void SaveChatLog(string userMessage, string botResponse, int index)
    {
        string timestamp = DateTime.Now.ToString("yyyy-MM-dd HH:mm:ss");
        var namingConvention = CamelCaseNamingConvention.Instance;
        List<ChatLogEntry> entries = [];

        if (File.Exists(ChatLogYamlFilePath))
        {
            string yamlContent = File.ReadAllText(ChatLogYamlFilePath);
            var deserializer = new DeserializerBuilder().WithNamingConvention(namingConvention).Build();
            entries = deserializer.Deserialize<List<ChatLogEntry>>(yamlContent);
        }
        entries ??= [];
        entries.Add(new ChatLogEntry()
        {
            Index = index,
            Timestamp = timestamp,
            UserMessage = userMessage,
            ChatbotResponse = botResponse
        });

        var serializer = new SerializerBuilder().WithNamingConvention(namingConvention).Build();
        File.WriteAllText(ChatLogYamlFilePath, serializer.Serialize(entries), Encoding.UTF8);
    }
}